src/middlewared/middlewared/plugins/vm.py

```diff
@@ -810,39 +811,85 @@ def activate_sharefs(self, pool_name=None):
 
         return memory_allocation
 
-    async def __set_guest_vmemory(self, memory):
-        usermem = sysctl.filter('hw.usermem')
-        max_arc = sysctl.filter('vfs.zfs.arc_max')
-        guest_mem_used = await self.get_vmemory_in_use()
-        memory = memory * 1024 * 1024
-
-        # Keep at least 35% of memory from initial arc_max.
-        throttled_arc_max = int(usermem[0].value * 1.35) - usermem[0].value
-        # Get the user memory and keep space for ARC.
-        throttled_user_mem = int(usermem[0].value - throttled_arc_max)
-        # Potential memory used by guests.
-        memory_used = guest_mem_used['RPRD'] + guest_mem_used['RNP']
-
-        vms_memory = memory_used + memory
-        if vms_memory <= throttled_user_mem:
-            if max_arc[0].value > throttled_arc_max:
-                if max(max_arc[0].value - memory, 0) != 0:
-                    self.logger.info('===> Setting ARC FROM: {} TO: {}'.format(max_arc[0].value, max_arc[0].value - memory))
-                    sysctl.filter('vfs.zfs.arc_max')[0].value = max_arc[0].value - memory
-            return True
-        else:
+    @accepts(Bool('overcommit', default=False))
+    def get_available_memory(self, overcommit):
+        """
+        Get the current maximum amount of available memory to be allocated for VMs.
+
+        If `overcommit` is true only the current used memory of running VMs will be accounted for.
+        If false all memory (including unused) of runnings VMs will be accounted for.
+
+        This will include memory shrinking ZFS ARC to the minimum.
+
+        Memory is of course a very "volatile" resource, values may change abruptly between a
+        second but I deem it good enough to give the user a clue about how much memory is
+        available at the current moment and if a VM should be allowed to be launched.
+        """
+        # Use 90% of free memory to play safe
+        free = int(psutil.virtual_memory().free * 0.9)
```

Comment:
Why `free` instead of `available`? It does not include `inactive` memory which still can be used.

Reply:
Nice finding ; )